vsphere/virtual_machine_config_structure.go

```diff
@@ -0,0 +1,664 @@
+		AfterResume:         getBoolWithRestart(d, "run_tools_scripts_after_resume"),
+		BeforeGuestStandby:  getBoolWithRestart(d, "run_tools_scripts_before_guest_standby"),
+		BeforeGuestShutdown: getBoolWithRestart(d, "run_tools_scripts_before_guest_shutdown"),
+		BeforeGuestReboot:   getBoolWithRestart(d, "run_tools_scripts_before_guest_reboot"),
+	}
+	return obj
+}
+
+// flattenToolsConfigInfo reads various fields from a
+// ToolsConfigInfo into the passed in ResourceData.
+func flattenToolsConfigInfo(d *schema.ResourceData, obj *types.ToolsConfigInfo) error {
+	d.Set("sync_time_with_host", obj.SyncTimeWithHost)
+	d.Set("run_tools_scripts_after_power_on", obj.AfterPowerOn)
+	d.Set("run_tools_scripts_after_resume", obj.AfterResume)
+	d.Set("run_tools_scripts_before_guest_standby", obj.BeforeGuestStandby)
+	d.Set("run_tools_scripts_before_guest_shutdown", obj.BeforeGuestShutdown)
+	d.Set("run_tools_scripts_before_guest_reboot", obj.BeforeGuestReboot)
+	return nil
+}
+
+// schemaVirtualMachineResourceAllocation returns the respective schema keys
+// for the various kinds of resource allocation settings available to a virtual
+// machine. This is an abridged version of ResourceAllocationInfo with only the
+// keys present that make sense for virtual machines.
+func schemaVirtualMachineResourceAllocation() map[string]*schema.Schema {
+	s := make(map[string]*schema.Schema)
+	shareLevelFmt := "The allocation level for %s resources. Can be one of high, low, normal, or custom."
+	shareCountFmt := "The amount of shares to allocate to %s for a custom share level."
+	limitFmt := "The maximum amount of memory (in MB) or CPU (in MHz) that this virtual machine can consume, regardless of available resources."
+	reservationFmt := "The amount of memory (in MB) or CPU (in MHz) that this virtual machine is guaranteed."
+
+	for _, t := range virtualMachineResourceAllocationTypeValues {
+		shareLevelKey := fmt.Sprintf("%s_share_level", t)
+		shareCountKey := fmt.Sprintf("%s_share_count", t)
+		limitKey := fmt.Sprintf("%s_limit", t)
+		reservationKey := fmt.Sprintf("%s_reservation", t)
+
+		s[shareLevelKey] = &schema.Schema{
+			Type:         schema.TypeString,
+			Optional:     true,
+			Default:      string(types.SharesLevelNormal),
+			Description:  fmt.Sprintf(shareLevelFmt, t),
+			ValidateFunc: validation.StringInSlice(sharesLevelAllowedValues, false),
+		}
+		s[shareCountKey] = &schema.Schema{
+			Type:         schema.TypeInt,
+			Optional:     true,
+			Computed:     true,
+			Description:  fmt.Sprintf(shareCountFmt, t),
+			ValidateFunc: validation.IntAtLeast(0),
+		}
+		s[limitKey] = &schema.Schema{
+			Type:         schema.TypeInt,
+			Optional:     true,
+			Default:      -1,
+			Description:  fmt.Sprintf(limitFmt, t),
+			ValidateFunc: validation.IntAtLeast(-1),
+		}
+		s[reservationKey] = &schema.Schema{
+			Type:         schema.TypeInt,
+			Optional:     true,
+			Default:      0,
```

Comment:
_minor_ I don't think this is needed, since it's defaulted to 0?

Reply:
Changing this to be consistent, but as mentioned previously, might end up adding a lot more zero values given what I encountered when tainting, although it doesn't seem to affect this stuff here. Might be part of a larger maintenance effort later on.